homeassistant/components/growatt_server/sensor.py

```diff
@@ -0,0 +1,161 @@
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon of the sensor."""
+        return "mdi:solar-power"
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def device_class(self):
+        """Return the device class of the sensor."""
+        return "power"
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return "W"
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the monitored installation."""
+        attributes = self.attributes
+        return attributes
+
+    def update(self):
+        """Get the latest data from the Growat API and updates the state."""
+
+        try:
+            self.api.login(self.username, self.password)
+            if self.is_total:
+                total_info = self.api.plant_info(self.inverter_id)
+                del total_info["deviceList"]
+                self._state = total_info["invTodayPpv"]
+                self.attributes = {
+                    "money_today": total_info["plantMoneyText"],
+                    "money_total": total_info["totalMoneyText"],
+                    "energy_today": total_info["todayEnergy"],
+                    "current_wattage": total_info["invTodayPpv"],
+                    "current_voltage": total_info["totalEnergy"],
+                    "maximum_output": total_info["nominalPower"],
+                }
+            else:
+                inverter_info = self.api.inverter_detail(self.inverter_id)
+
+                self._state = inverter_info["data"]["pac"]
+                self.attributes = {
+                    "energy_today": inverter_info["data"]["e_today"],
+                    "energy_total": inverter_info["data"]["e_total"],
+                    "voltage_input_1": inverter_info["data"]["vpv1"],
+                    "amperage_input_1": inverter_info["data"]["ipv1"],
+                    "wattage_input_1": inverter_info["data"]["ppv1"],
+                    "voltage_input_2": inverter_info["data"]["vpv2"],
+                    "amperage_input_2": inverter_info["data"]["ipv2"],
+                    "wattage_input_2": inverter_info["data"]["ppv2"],
+                    "voltage_input_3": inverter_info["data"]["vpv3"],
+                    "amperage_input_3": inverter_info["data"]["ipv3"],
+                    "wattage_input_3": inverter_info["data"]["ppv3"],
+                    "inverter_internal_wattage": inverter_info["data"]["ppv"],
+                    "inverter_reactive_voltage": inverter_info["data"]["vacr"],
+                    "inverter_reactive_amperage": inverter_info["data"]["iacr"],
+                    "frequency": inverter_info["data"]["fac"],
+                    "current_wattage": inverter_info["data"]["pac"],
+                    "current_reactive_wattage": inverter_info["data"]["pacr"],
```

Comment:
I suggest we instead add a sensor per each of these attributes. We prefer many individual sensors instead of many attributes on a few sensors.

Reply:
do you have a good example of how to update these sensors without making a request for every single sensor. since right now it only creates 3 sensors in my case which all make their own request since the data is different. however how would i get the correct data to the right sensor without making a request for every single sensor.

Reply:
Make a data class which instance is passed to each sensor. The data class is responsible for fetching new data and caching it. The update method in the data class is throttled. The entities poll the data instance update method and get the correct state value from the data instance state cache.

Reply:
Thank you, i've managed to successfully implement this.
However i'm getting quite overwhelmed by the amount of sensors being added.
Is there a good way to include or exclude specific sensors from being added and stop them from being tracked.

Reply:
If we implement the `unique_id` property on the entities the user will be able to disable entities. We can then also disable entities by default in the integration.

https://developers.home-assistant.io/docs/en/entity_index.html#generic-properties

https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirementshttps://developers.home-assistant.io/docs/en/entity_registry_disabled_by.html

Maybe use the plant_id and the sensor type as unique_id? Is the plant_id unique and stable?

Reply:
The `plant_id` should always be stable and linked to the same inverters.
That sounds like a good id, i'm thinking of using the inverter deviceSn and sensor type as unique id for that.

Reply:
Great!

Reply:
I added the unique id and it shows up in the entity registry :smile: